pkg/session_manager/session_manager.go

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2023 SentinelOne, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package session_manager
+
+import (
+	"go.uber.org/zap"
+)
+
+type (
+	Func          func(string) (interface{}, error)
+	EventCallback func(key string, eventsChannel <-chan interface{})
+)
+
+// command represents command used to communicate within the session manager
+type command struct {
+	op    string
+	key   string
+	value interface{}
+}
+
+// SessionManager is a manager for session channels
+// It allows to subscribe, publish and unsubscribe to/from channels
+type SessionManager struct {
+	eventCallback EventCallback
+	channels      map[string]chan interface{}
+	logger        *zap.Logger
+	operations    chan command
+}
+
+// New creates a new SessionManager
+func New(
+	logger *zap.Logger,
+	eventsCallback EventCallback,
+) *SessionManager {
+	manager := &SessionManager{
+		eventCallback: eventsCallback,
+		channels:      make(map[string]chan interface{}),
+		logger:        logger,
+		operations:    make(chan command),
+	}
+
+	go manager.processCommands()
+
+	return manager
+}
+
+// Sub subscribes to a channel key
+// It's fine to call this function multiple times with the same key - only one
+// channel will be created
+func (manager *SessionManager) Sub(key string) {
+	manager.operations <- command{op: "sub", key: key}
+}
+
+func (manager *SessionManager) sub(key string) {
+	_, found := manager.channels[key]
+	if !found {
+		manager.logger.Debug("Subscribing to key", zap.String("key", key))
+		ch := make(chan interface{})
+		manager.channels[key] = ch
+		go manager.eventCallback(key, ch)
+	}
+}
+
+// Pub publishes a value to a channel key
+func (manager *SessionManager) Pub(key string, value interface{}) {
+	manager.operations <- command{op: "pub", key: key, value: value}
+}
+
+func (manager *SessionManager) pub(key string, value interface{}) {
```

Comment:
Please, don't use this function naming - it is very confusing. Technically, you could have a public function named Pub and a private function named pub in the same package. However, this could still be confusing for anyone reading or maintaining the code.